docs/adapter_impl_notes.md

```diff
@@ -0,0 +1,184 @@
+
+The main function exported by ``gmail-adapter.js`` takes a ``config`` argument containing the configuration object for the adapter, and returns an object with ``name`` and ``read`` attributes. The Gmail adapter only implements ``read`` and not ``write``. Adapters that additionally support ``write`` would have an additional attribute ``write``. THe value for the ``name``' attribute is ``gmail``, corresponding to the ``read gmail`` proc in juttle programs. The value for ``read``' is an object inheriting from ``Juttle.proc.base``, which performs the work of the adapter.
+
+Here's the exported function from ``gmail-backend.js``:
+
+```Javascript
+var GmailBackend = function(config) {
+
+    auth = authorize(config["client-credentials"],
+                     config["oauth2-token"]);
+
+    return {
+        name: 'gmail',
+        read: Read
+    };
+};
+
+module.exports = GmailBackend;
+```
+
+The ``Read`` object should implement the following methods:
+
+```
+initialize: function(options, params, pname, location, program, juttle) {...}
+```
+
+``initialize`` is called when the program is compiled, and is responsible for validating and parsing the parameters and filter expression. The arguments to ``initialize`` are the following:
+
+* ``options``: The options provided to the ``read gmail`` proc in the juttle program. ``-from``, ``-to``, etc. are all examples of options.
+* ``params``: The filter expression is passed as ``params.filter_ast``.
+* ``pname``: XXX/mstemm is this needed?
+* ``location``: The location in the juttle program where this proc is located. It has been saved to ``this.location`` by the parent class.
+* ``program``: A reference to the associated ``Program`` object.
+* ``juttle``: XXX/mstemm is this needed?
+
+```
+start: function() {...}
+```
+
+```start``` is called when the program starts. At this time the adapter fetches the relevant messages given the search expression, constructs points from the messages, and calls ``emit()`` to pass the points to the program.
+
+```
+teardown: function() {...}
+```
+
+``teardown`` is called when the program ends. At this time the adapter performs any necessary cleanup (not necessary for the gmail adapter).
+
+## Returning Errors
+
+``Juttle.proc.base`` provides a method ``compile_error(code, info)`` that can be
+called when there is an error in ``initialize``. Here's an example:
+
+```Javascript
+var unknown = _.difference(_.keys(options), this.allowed_options);
+if (unknown.length > 0) {
+    throw this.compile_error('RT-UNKNOWN-OPTION-ERROR', {
+        option: unknown[0]
+    });
+}
+```
+
+Code is one of the values in [juttle-error-strings-en-US.json](https://github.com/juttle/juttle/blob/master/lib/strings/juttle-error-strings-en-US.json) from the juttle repository. Based on the error code, the ``info``` object is used to interpolate the error string template into a specific error string.
```

Comment:
ah, there is the extra backtick, after info! that's why markdown is confused.

btw, the inlined code block is good enough between single backticks `like this`.